File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 25;
int prv[N];
vector<int> a;
int t, ans;

void dfs(int idx)
{
    if (idx == a.size())
    {
        int sum = 0;
        for (int i = 0; i < a.size(); i++)
        {
            if (prv[i] == 0) sum += a[i];
            else sum -= a[i];
        }
        
        if (sum == t)
            ans++;
        
        return;
    }
    for (int i = 0; i < 2; i++)
    {
        prv[idx] = i;
        dfs(idx + 1);
    }
}

int solution(vector<int> numbers, int target) 
{
    a = numbers; t = target;
    dfs(0);
    
    return ans;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution 
{
    int ans = 0;
    int t = 0;
    ArrayList<Integer> a = new ArrayList<>();
    
    public void dfs(int i, int sum)
    {
        if (i == a.size())
        {
            if (sum == t)
                ans++;
            return;
        }
        
        dfs(i + 1, sum + a.get(i));
        dfs(i + 1, sum - a.get(i));
    }
    
    public int solution(int[] v, int target) {
        t = target;
        for (int to : v) a.add(to);
        dfs(0, 0);
        return ans;
    }
}